plutus-metatheory/src/VerifiedCompilation/UFloatDelay.lagda.md

```diff
@@ -0,0 +1,80 @@
+---
+title: VerifiedCompilation.UFloatDelay
+layout: page
+---
+
+# Float-Delay Translation Phase
+```
+module VerifiedCompilation.UFloatDelay where
+
+```
+## Imports
+
+```
+open import VerifiedCompilation.Equality using (DecEq; _≟_;decPointwise)
+open import VerifiedCompilation.UntypedViews using (Pred; isCase?; isApp?; isLambda?; isForce?; isBuiltin?; isConstr?; isDelay?; isTerm?; allTerms?; iscase; isapp; islambda; isforce; isbuiltin; isconstr; isterm; allterms; isdelay)
+open import VerifiedCompilation.UntypedTranslation using (Translation; translation?; Relation; convert; reflexive)
+open import Relation.Nullary.Product using (_×-dec_)
+open import Untyped using (_⊢; case; builtin; _·_; force; `; ƛ; delay; con; constr; error)
+import Relation.Binary.PropositionalEquality as Eq
+open Eq using (_≡_; refl)
+open import Data.Empty using (⊥)
+open import Function using (case_of_)
+open import Agda.Builtin.Maybe using (Maybe; just; nothing)
+open import Data.List using (map)
+open import Relation.Nullary using (Dec; yes; no; ¬_)
+import Relation.Binary as Binary using (Decidable)
+open import Data.Product using (_,_)
+
+```
+## Translation Relation
+
+This translation "floats" delays in applied terms into the abstraction, without inlining the whole term.
+This is separate from inlining because the added `delay` may make it seem like a substantial increase in code
+size, and so dissuade the inliner from executing. However, it may be that the laziness inside the resulting term
+is more computationally efficient.
+
+This requires a function to substitute the delay into an abstraction without doing full β-substitution.
+```
+
+{-# TERMINATING #-}
+subs-delay : {X : Set}{{de : DecEq X}} → (v : Maybe X) → (Maybe X ⊢) → (Maybe X ⊢)
+subs-delay v (` x) with v ≟ x
+... | yes refl = (delay (` x))
+... | no _ = (` x)
+subs-delay v (ƛ t) = ƛ (subs-delay (just v) t) -- The de Brujin index has to be incremented
+subs-delay v (t · t₁) = (subs-delay v t) · (subs-delay v t₁)
+subs-delay v (force t) = force (subs-delay v t) -- This doesn't immediately apply Force-Delay
+subs-delay v (delay t) = delay (subs-delay v t)
+subs-delay v (con x) = con x
+subs-delay v (constr i xs) = constr i (map (subs-delay v) xs)
+subs-delay v (case t ts) = case (subs-delay v t) (map (subs-delay v) ts)
+subs-delay v (builtin b) = builtin b
+subs-delay v error = error
+
+```
+The translation relation is then fairly striaghtforward.
+
+```
+data FlD {X : Set} {{de : DecEq X}} : (X ⊢) → (X ⊢) → Set₁ where
+  floatdelay : {y y' : X ⊢} {x x' : (Maybe X) ⊢}
+          → Translation FlD (subs-delay nothing x) x'
+          → Translation FlD y y'
+          → FlD (ƛ x · (delay y)) (ƛ x' · y')
```

Comment:
So I've only attempted to make this thing type check without trying to think about its semantics, but now that I review the PR, this isn't a semantics-preserving transformation, is it?

If I'm reading it correctly, it'll transform `(\x z -> x) (delay y)` into `(\x z -> delay x) y`, but these are very different terms: the former doesn't necessarily force `y` while the latter does always force it.

I don't know what the pass does in Haskell, but what's written here doesn't look correct to me. I may be missing something though.

Reply:
Well, each pass needs to make sense on its own, so there should either be an explicit dependency of the force-delay one or something else that constraints this pass to only work over appropriate terms... except that wouldn't help either, because even if it was `force x`, you'd still evaluate `y` with the second term without necessarily evaluating it with the first one.

Reply:
I was also assuming that converting `(\x -> blah x blah x blah) . (delay y)` into `(\x -> blah (delay x) blah (delay x) blah) . y` provides more "opportunities for laziness", since the long term with lots of `x`s might realise that it doesn't need to evaluate all of them? That is more likely if they are at multiple places in a larger expression? Doing the `y` once, even if you always do it, then seems cheaper? Whether this is actually a correct impression of how the evaluation will proceed, though, I don't know?

Reply:
> Doing the `y` once, even if you always do it, then seems cheaper?

That's just not how things work, at all. `y` can be a loop or an error crashing the program, you can't change the semantics of a completely arbitrary term.

> (\x -> blah (delay x) blah (delay x) blah) . y provides more "opportunities for laziness", since the long term with lots of xs might realise that it doesn't need to evaluate all of them?

`x` is already evaluated within that term, since it's bound by a lambda. So evaluating `force (delay x)` amount to simply discharging the force-delay pair and looking `x` up in the current environment.

Reply:
I'm just playing "Devil's Advocate" :) This phase was created for _some_ purpose, we ought to find out what and document it...

Reply:
I think when we spoke about this transformation I glossed over some details, sorry. I've looked at the docs again, and it's _necessary_ that every occurrence of the variable is immediately under a `force`, and that the argument is ["work-free"](https://github.com/IntersectMBO/plutus/blob/e2016d5ee6655003ea7341ed26fd76328e1105b3/plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/FloatDelay.hs#L131).

So, let's take a look at some examples. Let `x, y, z ...` be variables and `arg` be some term; we assume that there aren't any occurrences of `x` in the body other than the ones explicitly mentioned:
1. `(\x -> ... force x ... force x ... x ... force x) (delay arg)` here we _cannot_ apply the transformation, because the third `x` in the body is not under a `force`.
2. `(\x -> ... force x ... force x ... force x ... force x) (delay y)`, although all `x`'s are under `force`, we still can't apply the transformation because `y` is not "work-free".
3. `(\x -> ... force x ... force x ... force x ... force x ...) (delay (\y -> ...))` in this case we can apply the transformation and the term becomes `(\x -> ... force (delay x) ... force (delay x) ... force (delay x) ... force (delay x) ...) (\y -> ...)`

I think you asked me what happens in the following case: `(\x -> (\y -> blah y blah) . x) . (delay z)`. Given the restrictions above, the transformation wouldn't apply, but if we had instead `(\x -> (\y -> blah y blah) . (force x)) . (delay arg)` where `arg` is "work-free", then the result would be `(\x -> (\y -> blah y blah) . (force (delay x))) . arg`.

Regarding its relation to the `force-delay` transformation, it looks like `float-delay` provides more opportunities for `force-delay` to fire, and `force-delay` provides more opportunities for `float-delay` to fire as well. If you consider a term like `(\x -> ... (force ... (force x)) ...) (delay ... (delay arg))` where `arg` is "work-free", and the number of forces over `x` and delays over `arg` is equal, one application of `float-delay` would "push" a single `delay` under the `force`. It's necessary to then run `force-delay` to simplify the `force (delay x)` pair to `x`, and then you can apply `float-delay` again. So depending on the number of simplifier iterations, a number of these force-delay pairs can get simplified away. In practice, I think the simplifier iterates about 3 times which seems to be enough for real-life programs.

If you ignore the restrictions, you get an unsound transformation. For example, if you apply it even if not every `x` is under a `force`:
```
(\x -> (\y -> x) . (force x)) . (delay arg)
--float-delay-->
(\x -> (\y -> x) . (force (delay x)) . arg
--beta-reduction-->
(\x -> x) . arg
--beta-reduction-->
arg

(\x -> (\y -> x) . (force x)) . (delay arg)
--beta-reduction-->
(\x -> x) . (delay arg)
--beta-reduction-->
delay arg

arg =/= delay arg
```

I am not entirely sure, however, if the "work-free" requirement is mostly to do with optimisation or if it also protects against modifying the semantics of impure terms. I'll have to look into that.

Reply:
I think when we spoke about this transformation I glossed over some details, sorry. I've looked at the docs again, and it's _necessary_ that every occurrence of the variable is immediately under a `force`, and that the argument is ["work-free"](https://github.com/IntersectMBO/plutus/blob/e2016d5ee6655003ea7341ed26fd76328e1105b3/plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/FloatDelay.hs#L131).

So, let's take a look at some examples. Let `x, y, z ...` be variables and `arg` be some term; we assume that there aren't any occurrences of `x` in the body other than the ones explicitly mentioned:
1. `(\x -> ... force x ... force x ... x ... force x) (delay arg)` here we _cannot_ apply the transformation, because the third `x` in the body is not under a `force`.
2. `(\x -> ... force x ... force x ... force x ... force x) (delay y)`, although all `x`'s are under `force`, we still can't apply the transformation because `y` is not "work-free".
3. `(\x -> ... force x ... force x ... force x ... force x ...) (delay (\y -> ...))` in this case we can apply the transformation and the term becomes `(\x -> ... force (delay x) ... force (delay x) ... force (delay x) ... force (delay x) ...) (\y -> ...)`

I think you asked me what happens in the following case: `(\x -> (\y -> blah y blah) . x) . (delay z)`. Given the restrictions above, the transformation wouldn't apply, but if we had instead `(\x -> (\y -> blah y blah) . (force x)) . (delay arg)` where `arg` is "work-free", then the result would be `(\x -> (\y -> blah y blah) . (force (delay x))) . arg`.

Regarding its relation to the `force-delay` transformation, it looks like `float-delay` provides more opportunities for `force-delay` to fire, and `force-delay` provides more opportunities for `float-delay` to fire as well. If you consider a term like `(\x -> ... (force ... (force x)) ...) (delay ... (delay arg))` where `arg` is "work-free", and the number of forces over `x` and delays over `arg` is equal, one application of `float-delay` would "push" a single `delay` under the `force`. It's necessary to then run `force-delay` to simplify the `force (delay x)` pair to `x`, and then you can apply `float-delay` again. So depending on the number of simplifier iterations, a number of these force-delay pairs can get simplified away. In practice, I think the simplifier iterates about 3 times which seems to be enough for real-life programs.

If you ignore the restrictions, you get an unsound transformation. For example, if you apply it even if not every `x` is under a `force`:
```
(\x -> (\y -> x) . (force x)) . (delay arg)
--float-delay-->
(\x -> (\y -> x) . (force (delay x)) . arg
--beta-reduction-->
(\x -> x) . arg
--beta-reduction-->
arg

(\x -> (\y -> x) . (force x)) . (delay arg)
--beta-reduction-->
(\x -> x) . (delay arg)
--beta-reduction-->
delay arg

arg =/= delay arg
```

I am not entirely sure, however, if the "work-free" requirement is mostly to do with optimisation or if it also protects against modifying the semantics of impure terms. I'll have to look into that.